Various Improvements

jdlr-au opened the pull request:
Got a nice opportunity to sit down and properly organise some things I've hacked on before.
Also took the opportunity to collect some commonly-requested features and bugfixes (thanks to all who contributed fixes via PRs, even if I didn't use the exact same solutions).

Draft only for now to collect feedback (especially around custom sensor units and default column width).
Also need to update the readme with some of the new options still.

I recommend reviewing each commit individually for more focused changes and manageable diff sizes.

General layout improvements:
- Avoid wind forecast and graph falling out of alignment when columns too narrow
- Set lower bound on minimum column width that results in "nice" (subjective, but everything fitting in most locales) layout
  - Can be overridden manually in settings page if not to user satisfaction
- The forecast can now scroll horizontally when not enough space to display all columns

Features:
- Options to hide forecast precipitation unit, wind unit, and hourly date label (at midnight)
- Conversion to/from more wind speed and pressure units (new options not exposed in UI yet)
- Can set unit for custom sensors (might be a better way to do this? not sure)
- Precipitation graph supports cm unit

Fixes:
- Avoid name conflict with built-in HA card `visibility` (renamed entity option to `visibility_entity`, retained ability to load old config)
- "NNW" and "WNW" wind directions no longer have icon swapped
- Autoscroll affects all forecast items (not just the main chart), and no longer tries to call non-existent function
- Always use the unit of weather entity/custom sensor, don't use system units
  - May not match in some cases, and avoids bad assumptions like length unit exactly specifying precipitation unit
- Wind gust speed unit is converted (same as wind speed)
- Wind speed used to sometimes not load correctly
- Precipitation in inches now uses smaller scale for hourly chart (like mm does)
- Feels like temperature is no longer hidden when equal to 0

Cleanup:
- Improved state management (basically just using Lit's update mechanisms and properties more appropriately)
  - More efficient (fewer unnecessary re-renders)
  - Should also be less likely to somehow enter improper states (if possible)
- Unit conversions re-implemented by converting to metric and back (much simpler)

jdlr-au committed:
horizontal scrolling, associated layout improvements

- make chart scrollable
  - addition of .scroll-content container
    (allows contents to scroll)
  - min-width styles on .chart-container, .conditions, and .wind-details
    (so not limited by card width if need more space to show forecasts)
- calculate minimum width of a column (columnMinWidth)
  - used to set appropriate min-width for scrolling containers
  - also used for numberOfForecasts == 0
  - rough calculation, based on label font size
    (maybe in future, can actually measure text width)
- improve layout for narrow columns
  - wind speed on separate line to direction symbol
  - removed whitespace between wind speed and unit
- account for padding (and margin?) when determining number of forecasts
  - (when numberOfForecasts == 0)

jdlr-au committed:
fix swapped wind direction icon for "NNW" and "WNW" directions

jdlr-au committed:
rename visibility entity option to "visibility_entity" (should fix #280)

this avoids the name conflicting with Home Assistant's new visibility option (as of 2024.6)

compatibility is retained for old configs by:
 1. if config.visibility_entity is set, use it
 2. otherwise, try to use config.visibility only if it is present in hass.states
    (will be undefined if it isn't set to an entity)

jdlr-au committed:
improve state management

just gets rid of some cruft and makes things work more predictably:
- call updateChart after Lit updated if *forecasts* change (not weather)
- don't use drawChart in forecast event callback
  - just let Lit do its thing based on modifying forecasts property
- autoscroll requests an update via Lit, instead of trying to update things
  itself (fixes #259)
- apply autoscroll directly to forecasts in Lit update (before render)
  - it now affects condition icons and wind details too
  - placement here avoids timing race conditions around the new hour
    (if performed separately in multiple places, it might be possible to
    sometimes get different results)
  - but side effect: even if autoscroll is later disabled, that data is
    permanently lost until next forecast event
- measureCard no longer has side effect of recreating chart
  - (even though that may be desired, leave it up to caller)

jdlr-au committed:
avoid unnecessary renders whenever _hass changes

instead make all weather properties are tracked by Lit

jdlr-au committed:
clean up and improve unit handling

- allow specifying units of sensors
  - for sensors that don't return values in same unit as weather
    (usually should be converted by Home Assistant, but not always)
  - falls back to weather entity unit if not set
- don't rely on system-wide unit settings
  - always get them from the sensor or weather entity
  - and especially don't use unrelated units (length for precipitation?)
  - many things already depended on units from the weather entity anyway
- simplified unit conversions, support for converting more units
- wind gust speed is also converted
- precipitation graph supports cm unit, and fixed scale for hourly inches

jdlr-au committed:
fix wind and pressure sometimes not loading properly

was caused by the conversion units falling back to units from this.weather
before this.weather had been set

jdlr-au committed:
use NaN check instead of truthy check to show roundedFeelsLike

jdlr-au committed:
reduce forecast layout issues when localised, add options for denser layout

- wind unit can wrap to second line
- column width is wide enough to display hourly forecast dates more often
  - mainly affects locales that use long times or month names
- added options to hide dates, wind unit, and precipitation unit
  - column width will automatically shrink as these are disabled
- can manually set column width, if not happy with the new sizing

thetoph commented:
Thanks for taking time to do this. Hopefully, this gets merged, but seems like the original package might be unmaintained. In the meantime I've installed the version with your fixes. 🙏🏻

jdlr-au commented:
Yeah, it's a shame that maintenance is slow. Although of course maintaining a project is a big undertaking, so it's very understandable.
If it were quick and easy, I'd consider forking, but I'd rather not dedicate a ton of time to it rn.

greendog99 commented:
Thanks for this effort!

If not already addressed, you may want to include a fix for the time display memory leak (setInterval() never being cleared resulting in massive memory leak) per #303.

jdlr-au committed:
better manage updateClock interval and autoscroll timeout

- updateClock and its associated interval are no longer a side-effect of
    renderMain. Instead, updateClock is moved to a separate method and
    setInterval is called in firstUpdated/updated as necessary.
- Added _updateClockInterval to track any existing interval for updateClock.
    setInterval will not be called when an interval has already been set, and
    any interval will be cleared when the clock is disabled or the card is
    disconnected.
- Autoscroll timeout is also cancelled when the card is disconnected.

Done because of #303. I decided not to use a longer interval when seconds are
hidden though, because the interval isn't synchronised to wall clock time.
Using a shorter value helps ensure updates aren't delayed.

jdlr-au commented:
I didn't have that one, thanks for pointing it out. Should be fixed here too now (slightly different, but similar implementation).

folfy commented:
Hey, I tested your changes, and I found an issue with having a second chart on the same page, which now won't always render (stays empty, first one renders fine). Also occurs sometimes when changing parameters, but haven't figured out how to rly fix this. Still better than breaking HA usability (#303) though.

The 'Set lower bound on minimum column width that results in "nice"' thing was a slightly unexpected major change, but setting the override_min_column_width allowed me to get back my original, non-scrolling 24h chart -> This option seems to be mssing in the readme (but found it later on in the UI editor).

Thanks for your work!
